# **ESCUELA POLITECNICA NACIONAL** 
### Nombre:Stiv Quishpe 
### Link al respositorio 


# Pregunta 1

Suponga que dos puntos \((x_0, y_0)\) y \((x_1, y_1)\) se encuentran en línea recta con \(y_1\) no es igual a \(y_0\).  
Existen dos fórmulas para encontrar la intersección \(x\) de la línea:

### Método A:
$$
x = \frac{y_0 x_1 - x_0 y_1}{y_1 - y_0}
$$

### Método B:
$$
x = x_0 - \frac{(x_1 - x_0) y_0}{y_1 - y_0}
$$

Usando los datos \((x_0, y_0) = (1.31, 3.24)\) y \((x_1, y_1) = (1.93, 4.76)\), determine el valor real de la intersección \(x\) (asumiendo redondeo a 6 cifras significativas):  

| \(X\)   | 1.31 | 1.93 |
|---------|------|------|  
| \(y\)   | 3.24 | 4.76 |

#### Método A:
$$
x = \frac{3.24 \cdot 1.93 - 1.31 \cdot 4.76}{4.76 - 3.24}
$$
$$
x = \frac{6.2532 - 6.2356}{1.52}
$$
$$
x = \frac{0.0176}{1.52} = 0.0115789
$$

#### Método B:
$$
x = 1.31 - \frac{(1.93 - 1.31) \cdot 3.24}{4.76 - 3.24}
$$
$$
x = 1.31 - \frac{0.62 \cdot 3.24}{1.52}
$$
$$
x = 1.31 - \frac{2.0088}{1.52}
$$
$$
x = 1.31 - 1.321578947 = 0.0115789
$$

Los dos métodos dan el mismo valor para \(x\).



### Usando aritmética de computadora con redondeo a 3 cifras significativas:
#### Método A:
$$
x = \frac{3.24 \cdot 1.93 - 1.31 \cdot 4.76}{4.76 - 3.24}
$$
$$
x = \frac{6.25 - 6.22}{1.52} = 0.0197
$$

**Error relativo:**  
$$
\text{error} = \frac{|0.0115789 - 0.0197|}{|0.0115789|} = 0.700
$$

#### Método B:
$$
x = 1.31 - \frac{(1.93 - 1.31) \cdot 3.24}{4.76 - 3.24}
$$
$$
x = 1.31 - 0.01 = 0.01
$$

**Error relativo:**  
$$
\text{error} = \frac{|0.0115789 - 0.01|}{|0.0115789|} = 0.136
$$

El **método B** es más preciso debido a que utiliza menos operaciones aritméticas.  



# Pregunta 2

Los primeros tres términos diferentes a cero de la serie de Maclaurin para la función arcotangente son:
$$
x - \frac{1}{3}x^3 + \frac{1}{5}x^5
$$

Calcule el error relativo en las siguientes aproximaciones de \(\pi\) mediante el polinomio (en lugar del arcotangente).  
Asuma que \(\pi = 3.14159\).  

## Aproximación 1: 4 · arctan(1/2) + 4 · arctan(1/3)


$$
\pi^* = 4\left(\left[\frac{1}{2} - \frac{1}{3}\left(\frac{1}{2}\right)^3 + \frac{1}{5}\left(\frac{1}{2}\right)^5\right] + \left[\frac{1}{3} - \frac{1}{3}\left(\frac{1}{3}\right)^3 + \frac{1}{5}\left(\frac{1}{3}\right)^5\right]\right)
$$
$$
\pi^* = 4\left(\frac{223}{480} + \frac{391}{1215}\right)
$$
$$
\pi^* = 4\left(\frac{6115}{7776}\right) = \frac{6115}{1944} = 3.1456
$$

**Error relativo:**
$$
\text{error} = \frac{|3.14159 - 3.1456|}{3.14159} = 0.0013
$$

## Aproximación 2: 16 · arctan(1/5) - 4 · arctan(1/239)


$$
\pi^* = 16\left(\frac{1}{5} - \frac{1}{3}\left(\frac{1}{5}\right)^3 + \frac{1}{5}\left(\frac{1}{5}\right)^5\right) - 4\left(\frac{1}{239} - \frac{1}{3}\left(\frac{1}{239}\right)^3 + \frac{1}{5}\left(\frac{1}{239}\right)^5\right)
$$
$$
\pi^* = 16\left(\frac{9253}{46875}\right) - 4(0.004184076)
$$
$$
\pi^* = 3.158357333 - 0.016736304 = 3.1416
$$

**Error relativo:**
$$
\text{error} = \frac{|3.14159 - 3.1416|}{3.14159} = 0.0000098
$$


# Pregunta 3


In [ ]:
def secant_method(f, x0, x1, tol=1e-6, max_iter=100):
  
    x_prev = x0
    x_curr = x1
    iter_count = 0
    f_x_prev = f(x_prev)
    f_x_curr = f(x_curr)

    while abs(f_x_curr) > tol and iter_count < max_iter:
        # Verifica que no haya división por cero
        if f_x_curr == f_x_prev:
            raise ValueError("División por cero: f(x_curr) y f(x_prev) son iguales.")
        
        # Calcula el siguiente punto
        x_next = x_curr - f_x_curr * (x_curr - x_prev) / (f_x_curr - f_x_prev)
        
        # Actualiza las variables para la próxima iteración
        x_prev, f_x_prev = x_curr, f_x_curr
        x_curr, f_x_curr = x_next, f(x_next)
        
        iter_count += 1

    if abs(f_x_curr) <= tol:
        return x_curr, iter_count
    else:
        raise ValueError("El método no convergió en el número máximo de iteraciones.")



# Ejemplo1

In [ ]:
i = 0

def func(x):
    global i
    i += 1
    y = x**3 - 3 * x**2 + x - 1
    print(f"Llamada i={i}\t x={x:.5f}\t y={y:.2f}")
    return y


secant_method(func, x0=2, x1=3)

Llamada i=1	 x=2.00000	 y=-3.00
Llamada i=2	 x=3.00000	 y=2.00
Llamada i=3	 x=2.60000	 y=-1.10
Llamada i=4	 x=2.74227	 y=-0.20
Llamada i=5	 x=2.77296	 y=0.03
Llamada i=6	 x=2.76922	 y=-0.00
Llamada i=7	 x=2.76929	 y=-0.00
Llamada i=8	 x=2.76929	 y=0.00


(2.7692923542484045, 6)

# Ejemplo2

In [ ]:
i = 0
import math

def func(x):
    global i
    i += 1
    y = math.sin(x) + 0.5
    print(f"Llamada i={i}\t x={x:.5f}\t y={y:.2f}")
    return y


secant_method(func, x0=2, x1=3)

Llamada i=1	 x=2.00000	 y=1.41
Llamada i=2	 x=3.00000	 y=0.64
Llamada i=3	 x=3.83460	 y=-0.14
Llamada i=4	 x=3.68602	 y=-0.02
Llamada i=5	 x=3.66399	 y=0.00
Llamada i=6	 x=3.66520	 y=-0.00
Llamada i=7	 x=3.66519	 y=-0.00


(3.66519143172732, 5)

# Problema 4


In [3]:
import math

def bisection_method(f, a, b, tol=1e-6, max_iter=100):
    if f(a) * f(b) > 0:
        raise ValueError("El intervalo no tiene cambio de signo.")

    iter_count = 0
    while (b - a) / 2 > tol and iter_count < max_iter:
        c = (a + b) / 2
        if f(c) == 0:  # Raíz exacta encontrada
            return c
        elif f(a) * f(c) < 0:
            b = c
        else:
            a = c
        iter_count += 1

    return (a + b) / 2

def f(x):
    return math.sin(x)

intervals = [
    (-1, 2),     
    (-5, 4),      
    (-2.5, -1),   
    (-4, 5),     
    (3, 5),       
    (-3.5, 3),    
]

for a, b in intervals:
    try:
        root = bisection_method(f, a, b)
        print(f"Intervalo [{a}, {b}]: Raíz encontrada ≈ {root}")
    except ValueError as e:
        print(f"Intervalo [{a}, {b}]: {e}")


Intervalo [-1, 2]: Raíz encontrada ≈ -2.384185791015625e-07
Intervalo [-5, 4]: Raíz encontrada ≈ -3.14159232378006
Intervalo [-2.5, -1]: El intervalo no tiene cambio de signo.
Intervalo [-4, 5]: Raíz encontrada ≈ 3.14159232378006
Intervalo [3, 5]: Raíz encontrada ≈ 3.1415929794311523
Intervalo [-3.5, 3]: El intervalo no tiene cambio de signo.


# Problema 5


In [ ]:
def newton_method(f, df, x0, tol=1e-6, max_iter=100):
    iter_count = 0
    x_curr = x0

    while iter_count < max_iter:
        f_x = f(x_curr)
        df_x = df(x_curr)
        
        if abs(df_x) < 1e-12:
            return "Error (división por 0)"

        x_next = x_curr - f_x / df_x
        
        if abs(x_next - x_curr) < tol:
            return x_next, iter_count + 1

        x_curr = x_next
        iter_count += 1

    return "Error (diverge u oscila)"

def f(x):
    return x**3 + x - (1 + 3*x**2)

def df(x):
    return 3*x**2 + 1 - 6*x

initial_values = [3, 1, 0, 1 + (6**0.5)/3]

for x0 in initial_values:
    result = newton_method(f, df, x0)
    print(f"x0 = {x0:.4f} -> Resultado: {result}")


x0 = 3.0000 -> Resultado: (2.7692923542387, 4)
x0 = 1.0000 -> Resultado: Error (diverge u oscila)
x0 = 0.0000 -> Resultado: Error (diverge u oscila)
x0 = 1.8165 -> Resultado: Error (división por 0)


# Resolviendo el problema matemáticamente usando el método de Newton-Raphson

El **método de Newton-Raphson** se basa en la siguiente fórmula:

$
x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}
$

En este caso, tenemos la función:

$
f(x) = x^3 + x - (1 + 3x^2)
$

## 1. Derivar la función f(x) 
Para aplicar el método de Newton, necesitamos la derivada de  f(x) :

$
f'(x) = 3x^2 + 1 - 6x
$

## 2. Procedimiento de iteración
Para calcular la raíz, iteramos usando la fórmula:

$
x_{n+1} = x_n - \frac{f(x_n)}{f'(x_n)}
$

donde $ x_0 $ es el valor inicial.

### Ejemplo de cálculos con diferentes valores iniciales:

1. **Si \( x_0 = 3 \):**
   - $ f(3) = 3^3 + 3 - (1 + 3 \cdot 3^2) = 27 + 3 - 28 = 2 $
   - $ f'(3) = 3 \cdot 3^2 + 1 - 6 \cdot 3 = 27 + 1 - 18 = 10 $
   - Primera iteración:
    $
     x_1 = 3 - \frac{2}{10} = 2.8
   $
   - Continuamos hasta que el error sea menor que la tolerancia $( \text{tol} = 10^{-6} )$.

2. **Si \( x_0 = 1 \):**
   - Calculamos \( f(1) \) y \( f'(1) \):
     $
     f(1) = 1^3 + 1 - (1 + 3 \cdot 1^2) = 1 + 1 - 4 = -2
     $
     $
     f'(1) = 3 \cdot 1^2 + 1 - 6 \cdot 1 = 3 + 1 - 6 = -2
     $
   - Primera iteración:
     $
     x_1 = 1 - \frac{-2}{-2} = 1
     $
   - El método no converge, ya que se repite el mismo valor.

3. **Si \( x_0 = 0 \):**
   - Calculamos \( f(0) \) y \( f'(0) \):
     $
     f(0) = 0^3 + 0 - (1 + 3 \cdot 0^2) = -1
     $
     $
     f'(0) = 3 \cdot 0^2 + 1 - 6 \cdot 0 = 1
     $
   - Primera iteración:
     $
     x_1 = 0 - \frac{-1}{1} = 1
     $
   - El método no converge, ya que oscila.

4. **Si $ x_0 = 1 + \sqrt{6}/3 $:**
   - f'(x)  en este punto es \( 0 \), lo que produce una división por cero. El método falla.

## 3. Conclusión
- Para $x_0 = 3 $, el método converge a $ x_{\text{sol}} = 2.76929 $.
- Para $ x_0 = 1$y $ x_0 = 0 $, el método diverge o oscila.
- Para $ x_0 = 1 + \sqrt{6}/3 $, ocurre un error de división por cero.


# Problema 6
## Resolución del Spline Cúbico Natural

### Problema:
Se deben calcular los coeficientes del spline cúbico \( S_0(x) \) para los puntos \((-1, 1)\) y \((1, 3)\), teniendo en cuenta las derivadas:
- \( f'(x_0) = 1 \)
- \( f'(x_n) = 2 \)

El spline cúbico tiene la forma general:

\[
S_0(x) = a(x - x_0)^3 + b(x - x_0)^2 + c(x - x_0) + d
\]

Donde los coeficientes \( a, b, c, d \) se determinan a partir de las condiciones dadas.

---

### 1. Condiciones Iniciales
Para \( S_0(x) \) que pase por \((-1, 1)\) y \((1, 3)\):

1. **Paso por el primer punto \((-1, 1)\):**
   \[
   S_0(-1) = 1
   \]
   Sustituyendo en la ecuación del spline:
   \[
   a(0)^3 + b(0)^2 + c(0) + d = 1
   \]
   Por lo tanto, \( d = 1 \)

2. **Condición en el segundo punto \((1, 3)\):**
   \[
   S_0(1) = 3
   \]
   Al sustituir en la ecuación del spline, obtenemos:
   \[
   a(2)^3 + b(2)^2 + c(2) + d = 3
   \]
   Sabemos que \( d = 1 \), por lo tanto:
   \[
   8a + 4b + 2c + 1 = 3
   \]
   Restando 1 de ambos lados:
   \[
   8a + 4b + 2c = 2
   \]

---

### 2. Derivadas en los extremos
1. **En el punto inicial (\( x = -1 \)):**
   \[
   S_0'(-1) = 1
   \]
   Derivando el spline:
   \[
   S_0'(x) = 3a(x - x_0)^2 + 2b(x - x_0) + c
   \]
   Evaluando en \( x = -1 \):
   \[
   S_0'(-1) = 3a(0)^2 + 2b(0) + c = 1
   \]
   Por lo tanto, \( c = 1 \)

2. **En el punto final (\( x = 1 \)):**
   \[
   S_0'(1) = 2
   \]
   Evaluando:
   \[
   S_0'(1) = 3a(2)^2 + 2b(2) + c
   \]
   Como \( c = 1 \):
   \[
   12a + 4b + 1 = 2
   \]
   Restando 1 de ambos lados:
   \[
   12a + 4b = 1
   \]

---
### 3. Resolviendo el sistema de ecuaciones
De las ecuaciones obtenidas:

1. **Ecuación 1:** `8a + 4b + 2 = 2`, lo que nos da `8a + 4b = 0`
2. **Ecuación 2:** `12a + 4b = 1`

Restamos la ecuación 1 de la ecuación 2:

`(12a + 4b) - (8a + 4b) = 1 - 0`

Esto nos da:

`4a = 1`

Por lo tanto, `a = 1/4`.

Sustituimos `a = 1/4` en la ecuación `8a + 4b = 0`:

`8 * (1/4) + 4b = 0`

Lo que simplifica a:

`2 + 4b = 0`

De aquí obtenemos:

`b = -1/2`.

---

### 4. Ecuación final del spline cúbico
Sustituyendo los coeficientes `a = 1/4`, `b = -1/2`, `c = 1`, `d = 1`, la ecuación final del spline cúbico es:

`S_0(x) = (1/4)(x + 1)^3 - (1/2)(x + 1)^2 + 1(x + 1) + 1`.


# Problema 7
# Interpolación de Lagrange

## Fórmula de Interpolación de Lagrange

La interpolación de un conjunto de puntos usando polinomios de Lagrange \( P(x) \) se expresa mediante la siguiente fórmula:

$
P(x) = \sum_{k=0}^{n} f(x_k) L_k(x)
$

Donde:

$
L_k(x) = \prod_{\substack{i=0 \\ i \neq k}}^{n} \frac{x - x_i}{x_k - x_i}
$

### Dado los puntos:
$
(0, 0), \quad (1, 1), \quad (2, 2), \quad (3, 3)
$

Para estos puntos, tenemos \( n = 3 \) ya que hay 4 puntos en total, y los valores de \( f(x_k) \) son iguales a \( x_k \), es decir, \( f(x_0) = 0 \), \( f(x_1) = 1 \), \( f(x_2) = 2 \), y \( f(x_3) = 3 \).

---

## Paso 1: Cálculo de los polinomios \( L_k(x) \)

Los polinomios $ L_k(x) $ se calculan de la siguiente manera:

1. Para $ k = 0 $:
   $
   L_0(x) = \frac{(x - 1)(x - 2)(x - 3)}{(0 - 1)(0 - 2)(0 - 3)} = \frac{(x - 1)(x - 2)(x - 3)}{-6}
   $
   Simplificando:
   $
   L_0(x) = \frac{-(x - 1)(x - 2)(x - 3)}{6}
   $

2. Para \( k = 1 \):
   $
   L_1(x) = \frac{(x - 0)(x - 2)(x - 3)}{(1 - 0)(1 - 2)(1 - 3)} = \frac{x(x - 2)(x - 3)}{2}
   $
   Simplificando:
   $
   L_1(x) = \frac{x(x - 2)(x - 3)}{2}
   $

3. Para  k = 2 :
   $
   L_2(x) = \frac{(x - 0)(x - 1)(x - 3)}{(2 - 0)(2 - 1)(2 - 3)} = \frac{x(x - 1)(x - 3)}{2}
   $
   Simplificando:
   $
   L_2(x) = \frac{x(x - 1)(x - 3)}{2}
   $

4. Para  k = 3 :
   $
   L_3(x) = \frac{(x - 0)(x - 1)(x - 2)}{(3 - 0)(3 - 1)(3 - 2)} = \frac{x(x - 1)(x - 2)}{6}
   $
   Simplificando:
   $
   L_3(x) = \frac{x(x - 1)(x - 2)}{6}
   $

---

## Paso 2: Construcción del Polinomio de Lagrange \( P(x) \)

El polinomio de interpolación de Lagrange es:

$
P(x) = f(x_0) L_0(x) + f(x_1) L_1(x) + f(x_2) L_2(x) + f(x_3) L_3(x)
$

Sustituyendo los valores \( f(x_k) \) en la fórmula:

$
P(x) = 0 \cdot L_0(x) + 1 \cdot L_1(x) + 2 \cdot L_2(x) + 3 \cdot L_3(x)
$

$
P(x) = L_1(x) + 2L_2(x) + 3L_3(x)
$
Sustituyendo las expresiones de \( L_1(x) \), \( L_2(x) \), y \( L_3(x) \):

$
P(x) = \frac{x(x - 2)(x - 3)}{2} + 2 \cdot \frac{x(x - 1)(x - 3)}{2} + 3 \cdot \frac{x(x - 1)(x - 2)}{6}
$

Simplificando:

$
P(x) = \frac{x(x - 2)(x - 3)}{2} + \frac{2x(x - 1)(x - 3)}{2} + \frac{3x(x - 1)(x - 2)}{6}
$

Multiplicando por 6 para eliminar denominadores:

$
P(x) = \frac{3x(x - 2)(x - 3)}{6} + 6x(x - 1)(x - 3) + x(x - 1)(x - 2)
$

Finalmente, combinamos y simplificamos los términos para obtener el polinomio completo.

---

## Paso 3: Evaluación del Polinomio

### 1. Evaluando P(3.78) :

Sustituimos x = 3.78  en el polinomio P(x) :

$
P(3.78) = \frac{3(3.78)(3.78 - 2)(3.78 - 3)}{6} + 6(3.78)(3.78 - 1)(3.78 - 3) + (3.78)(3.78 - 1)(3.78 - 2)
$

### 2. Evaluando  P(19.102) :

Sustituimos x = 19.102  en el polinomio  P(x) :

$
P(19.102) = \frac{3(19.102)(19.102 - 2)(19.102 - 3)}{6} + 6(19.102)(19.102 - 1)(19.102 - 3) + (19.102)(19.102 - 1)(19.102 - 2)
$



# Problema 8
## Resolución del Problema de Splines Cúbicos

### 1. Modificación de los Splines para Cumplir con la Pendiente Deseada \( m \) en el Punto \( (x_1, y_1) \):

Para cumplir con el requisito de la pendiente \( m \) en \( x_1 \), podemos modificar la condición de tangencia:

\[
S'_1(x_1) = m
\]

Esto significa que debemos ajustar los coeficientes \( b_0 \) y \( b_1 \) de los splines para que se cumpla la condición de la derivada en el punto \( x_1 \).

### 2. Condiciones Necesarias para la Existencia de una Solución:

Al trabajar con splines cúbicos, las condiciones para una solución única son las siguientes:

- **Condiciones de continuidad en las derivadas**:
  - \( S_1(x_1) = y_1 \)
  - \( S'_0(x_1) = S'_1(x_1) \) (continuidad de la primera derivada en \( x_1 \))
  - \( S''_0(x_0) = 0 \) (condición de frontera en \( x_0 \))
  - \( S''_1(x_2) = 0 \) (condición de frontera en \( x_2 \))

Si deseamos que la pendiente en \( x_1 \) sea \( m \), necesitamos modificar el sistema de ecuaciones para garantizar que \( S'_1(x_1) = m \). Por lo tanto, la ecuación que debe modificarse es la de la **continuidad de la primera derivada** en \( x_1 \):

\[
S'_0(x_1) = S'_1(x_1)
\]

### 3. Ecuaciones del Spline:

Dado los puntos \( (-1, 1) \), \( (0, 5) \), \( (1, 3) \), tenemos que encontrar los splines correspondientes.

#### Para \( S_0(x) \):

La forma general de \( S_0(x) \) es:

\[
S_0(x) = a_0 + b_0(x - x_0) + c_0(x - x_0)^2 + d_0(x - x_0)^3
\]

Sustituyendo \( x_0 = -1 \), \( y_0 = 1 \) y utilizando \( m = -3 \) (suponiendo del gráfico), obtenemos los coeficientes de \( S_0(x) \).

#### Para \( S_1(x) \):

La forma general de \( S_1(x) \) es:

\[
S_1(x) = a_1 + b_1(x - x_1) + c_1(x - x_1)^2 + d_1(x - x_1)^3
\]

Donde \( x_1 = 0 \), \( y_1 = 5 \), y la pendiente deseada \( m = -3 \).

### 4. Solución:

- **Spline \( S_0(x) \)**:

\[
S_0(x) = -3(x + 1)^2 + 7(x + 1) + 1
\]

- **Spline \( S_1(x) \)**:

\[
S_1(x) = -3x^2 + x + 5
\]

### Conclusión:

Para lograr la condición de pendiente deseada \( m \) en el punto \( x_1 \), es necesario modificar la continuidad de la primera derivada en el punto de unión \( x_1 \) y aplicar las condiciones correspondientes a los splines cúbicos. Las ecuaciones resultantes de los splines son:

- **Spline \( S_0(x) \)**:
  \[
  S_0(x) = -3(x + 1)^2 + 7(x + 1) + 1
  \]

- **Spline \( S_1(x) \)**:
  \[
  S_1(x) = -3x^2 + x + 5
  \]

